server/service.go

```diff
@@ -345,45 +348,53 @@ func (m *BoostService) handleGetHeader(w http.ResponseWriter, req *http.Request)
 			mu.Lock()
 			defer mu.Unlock()
 
+			// Remember which relays delivered which bids (multiple relays might deliver the top bid)
+			if _, ok := relays[blockHash]; !ok {
+				relays[blockHash] = []string{relay.String()}
+			} else {
+				relays[blockHash] = append(relays[blockHash], relay.String())
+			}
+
 			// Skip if value (fee) is not greater than the current highest value
 			if result.response.Data != nil && responsePayload.Data.Message.Value.Cmp(&result.response.Data.Message.Value) < 1 {
 				return
 			}
 
 			// Use this relay's response as mev-boost response because it's most profitable
-			log.Debug("received a better bid")
+			log.Debug("received a good bid")
 			result.response = *responsePayload
-			result.relay = relay.String()
+			result.blockHash = blockHash
 			result.t = time.Now()
 		}(relay)
 	}
 
 	// Wait for all requests to complete...
 	wg.Wait()
 
-	if result.response.Data == nil {
+	if result.blockHash == "" {
 		log.Info("no bid received")
 		w.WriteHeader(http.StatusNoContent)
 		return
 	}
 
+	// Return the bid
+	m.respondOK(w, result.response)
```

Comment:
responding to getHeader before updating `m.bids` has the drawback of a possible race with the subsequent getPayload
I don't think this is likely to happen in the wild, since the proposer has to sign the request but it's not out of the question

Reply:
indeed, perhaps better to keep it at the end